lib/concurrent/configuration.rb

```diff
@@ -1,16 +1,21 @@
 require 'thread'
-require 'concurrent/atomics'
 require 'concurrent/errors'
-require 'concurrent/executors'
 require 'concurrent/concern/deprecation'
 require 'concurrent/concern/logging'
-require 'concurrent/utility/at_exit'
-require 'concurrent/utility/processor_counter'
 
 module Concurrent
   extend Concern::Logging
   extend Concern::Deprecation
 
+  autoload :AtExit,             'concurrent/utility/at_exit'
+  autoload :AtomicReference,    'concurrent/atomic/atomic_reference'
+  autoload :Delay,              'concurrent/delay'
+  autoload :ProcessorCounter,   'concurrent/utility/processor_counter'
+  autoload :TimerSet,           'concurrent/executor/timer_set'
+  autoload :ImmediateExecutor,  'concurrent/executor/immediate_executor'
```

Comment:
The big problem is `Delay`. We use Delay extensively to lazy-load resources--especially in `configuration.rb`. Delay itself is relatively uncoupled from everything else. The rabbit-hole is `Executor.executor_from_options(opts)`. *Many* of our abstractions, including Delay, use that function to initialize an executor. That function, in turn, uses the global executor declarations from configuration.rb, most of which are lazy-loaded with Delay.

Since Delay doesn't need configuration data until an object is constructed we could potentially lazy-load configuration w/i Delay and possibly avoid the circular dependency.